src/main/java/org/prebid/server/bidder/apacdex/ApacdexBidder.java

```diff
@@ -0,0 +1,138 @@
+    }
+
+    private void modifyImp(Imp imp, ExtImpApacdex extImpApacdex) {
+        imp.toBuilder()
+                .ext(mapper.mapper().valueToTree(extImpApacdex))
+                .build();
+    }
+
+    private ExtImpApacdex parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), APACDEX_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage(), e);
+        }
+    }
+
+    private HttpRequest<BidRequest> makeHttpRequest(BidRequest bidRequest) {
+        return HttpRequest.<BidRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(endpointUrl)
+                .headers(HttpUtil.headers())
+                .payload(bidRequest)
+                .body(mapper.encodeToBytes(bidRequest))
+                .build();
+    }
+
+    @Override
+    public Result<List<BidderBid>> makeBids(HttpCall<BidRequest> httpCall, BidRequest bidRequest) {
+        final List<BidderError> errors = new ArrayList<>();
+        try {
+            final BidResponse bidResponse = mapper.decodeValue(httpCall.getResponse().getBody(), BidResponse.class);
+            return Result.of(extractBids(bidResponse, errors), errors);
+        } catch (DecodeException | PreBidException e) {
+            return Result.withError(BidderError.badServerResponse(e.getMessage()));
+        }
+    }
+
+    private List<BidderBid> extractBids(BidResponse bidResponse, List<BidderError> errors) {
+        if (bidResponse == null || CollectionUtils.isEmpty(bidResponse.getSeatbid())) {
+            return Collections.emptyList();
+        }
+
+        return bidResponse.getSeatbid().stream()
+                .filter(Objects::nonNull)
+                .map(SeatBid::getBid)
+                .filter(Objects::nonNull)
+                .flatMap(Collection::stream)
+                .filter(Objects::nonNull)
+                .map(bid -> constructBidderBid(bid, bidResponse, errors))
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+    }
+
+    private BidderBid constructBidderBid(Bid bid, BidResponse bidResponse, List<BidderError> errors) {
+        final JsonNode extNode = bid.getExt();
+        final JsonNode typeNode = isNotEmptyOrMissedNode(extNode) ? extNode.at("/prebid/type") : null;
+
+        if (typeNode == null || !typeNode.isTextual()) {
+            errors.add(BidderError.badInput("Unable to read bid.ext.prebid.type"));
+            return null;
+        }
+
+        return BidderBid.of(bid, getBidType(typeNode, bid.getImpid()), bidResponse.getCur());
+    }
+
+    private BidType getBidType(JsonNode typeNode, String impid) {
+        try {
+            return mapper.mapper().convertValue(typeNode, BidType.class);
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(String.format("Failed to parse bid media type for impression %s", impid));
+        }
+    }
```

Comment:
I think better to do it like this:
```
    private BidderBid constructBidderBid(Bid bid, String currency, List<BidderError> errors) {
        final JsonNode extNode = bid.getExt();
        final JsonNode typeNode = isNotEmptyOrMissedNode(extNode) ? extNode.at("/prebid/type") : null;

        try {
            return BidderBid.of(bid, getBidType(typeNode, bid.getImpid()), currency);
        } catch (PreBidException e) {
            errors.add(BidderError.badInput(e.getMessage()));
            return null;
        }
    }

    private BidType getBidType(JsonNode typeNode, String impid) {
        try {
            return mapper.mapper().convertValue(typeNode, BidType.class);
        } catch (IllegalArgumentException ignored) {
        }

        throw new PreBidException(String.format("Failed to parse bid media type for impression %s", impid));
    }
```

Reply:
we can omit `try{...} catch()` if we in `getBidType(...)` ignore exception

Reply:
i think we can leave `type Node` check on null